core/libp2p/transport/tcp/tcp_connection.hpp

```diff
@@ -8,64 +8,171 @@
+    }
+
+    outcome::result<void> connect(const Tcp::endpoint &endpoint) {
+      OUTCOME_TRY(resolved, resolve(endpoint));
+      initiator_ = true;
+      auto [ec, e] = boost::asio::async_connect(socket_, resolved, yield_);
+      if (ec) {
+        return handle_errcode(ec);
+      }
+      return outcome::success();
+    }
+
+    outcome::result<void> close() override {
+      boost::system::error_code ec;
+      socket_.close(ec);
+      if (ec) {
+        return handle_errcode(ec);
+      }
+
+      return outcome::success();
+    }
+
+    bool isClosed() const override {
+      return !socket_.is_open();
+    }
+
+    outcome::result<multi::Multiaddress> remoteMultiaddr() override {
+      try {
+        return detail::make_endpoint(socket_.remote_endpoint());
+      } catch (const boost::system::system_error &e) {
+        return handle_errcode(e.code());
+      };
+    }
+
+    outcome::result<multi::Multiaddress> localMultiaddr() override {
+      try {
+        return detail::make_endpoint(socket_.local_endpoint());
+      } catch (const boost::system::system_error &e) {
+        return handle_errcode(e.code());
+      };
+    }
+
+    outcome::result<std::vector<uint8_t>> read(size_t bytes) override {
+      std::vector<uint8_t> out(bytes, 0);
+      OUTCOME_TRY(read(out));
+      return out;
+    };
+
+    /**
+     * @brief Blocks until any number (not more that {@param bytes}) of bytes
+     * are read.
+     * @param bytes number of bytes to read
+     * @return vector of bytes or error code
+     */
+    outcome::result<std::vector<uint8_t>> readSome(size_t bytes) override {
+      std::vector<uint8_t> out(bytes, 0);
+      OUTCOME_TRY(size, readSome(out));
+      out.resize(size);
+      return out;
+    };
+
+    outcome::result<size_t> read(gsl::span<uint8_t> buf) override {
+      auto [ec, read] =
+          boost::asio::async_read(socket_, detail::make_buffer(buf), yield_);
+      if (ec) {
+        return handle_errcode(ec);
+      }
+
+      return read;
+    };
+
+    outcome::result<size_t> readSome(gsl::span<uint8_t> buf) override {
+      auto [ec, read] =
+          socket_.async_read_some(detail::make_buffer(buf), yield_);
+      if (ec) {
+        return handle_errcode(ec);
+      }
+
+      return read;
+    };
+
+    outcome::result<size_t> write(gsl::span<const uint8_t> in) override {
+      auto [ec, written] =
+          boost::asio::async_write(socket_, detail::make_buffer(in), yield_);
+      if (ec) {
+        return handle_errcode(ec);
+      }
+
+      return written;
+    };
+
+    outcome::result<size_t> writeSome(gsl::span<const uint8_t> in) override {
+      auto [ec, written] =
+          socket_.async_write_some(detail::make_buffer(in), yield_);
+      if (ec) {
+        return handle_errcode(ec);
+      }
+
+      return written;
+    };
+
+    bool isInitiator() const noexcept override {
+      return initiator_;
+    }
 
    private:
-    Tcp::resolver resolver_;
+    yield_t &yield_;
```

Comment:
Isn't it dangerous to keep such a type as a reference? Won't pointer be better?

Reply:
yield_t is allocated on stack and TcpConnection exists only within yield context, so lifetime is guaranteed to be as long as for TcpConnecion